File: 0502-ipo/0502-ipo.cs
public class Solution {
    public int FindMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.Length;
        
        List<(int capital, int profit)> projects = new List<(int, int)>();
        for (int i = 0; i < n; ++i) {
            projects.Add((capital[i], profits[i]));
        }
        
        projects.Sort((a, b) => a.capital.CompareTo(b.capital));
        
        PriorityQueue<int, int> maxProfitHeap = new PriorityQueue<int, int>(Comparer<int>.Create((a, b) =>              b.CompareTo(a)));
        
        int index = 0;
        for (int i = 0; i < k; ++i) {
            while (index < n && projects[index].capital <= w) {
                maxProfitHeap.Enqueue(projects[index].profit, projects[index].profit);
                ++index;
            }

            if (maxProfitHeap.Count == 0) {
                break;
            }

            w += maxProfitHeap.Dequeue();
        }
        
        return w;
    }
}

File: 0502-ipo/0502-ipo.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int n = profits.size();
        
        vector<pair<int, int>> projects;
        for (int i = 0; i < n; ++i) {
            projects.emplace_back(capital[i], profits[i]);
        }
        
        sort(projects.begin(), projects.end());
        
        priority_queue<int> maxProfitHeap;
        
        int index = 0;
        for (int i = 0; i < k; ++i) {
            while (index < n && projects[index].first <= w) {
                maxProfitHeap.push(projects[index].second);
                ++index;
            }
            
            if (maxProfitHeap.empty()) {
                break;
            }

            w += maxProfitHeap.top();
            maxProfitHeap.pop();
        }
        
        return w;
    }
};